File: workspaces/ws_001_tipcalcdeluxe/ios/TipCalculatorDeluxe/Models/CurrencyFormatter.swift
import Foundation

/// Locale-aware currency formatting. The currency symbol is auto-detected
/// from the device locale (spec: "Currency symbol auto-detection from locale").
enum CurrencyFormatter {

    /// Shared formatter configured for the current locale.
    private static let formatter: NumberFormatter = {
        let f = NumberFormatter()
        f.numberStyle = .currency
        f.locale = Locale.current
        f.maximumFractionDigits = 2
        f.minimumFractionDigits = 2
        return f
    }()

    /// The currency symbol for the current locale (e.g. "$", "£", "AED").
    static var symbol: String {
        formatter.currencySymbol ?? Locale.current.currency?.identifier ?? "$"
    }

    /// Format a monetary value using the current locale's currency style.
    static func string(from value: Double) -> String {
        let safe = value.isFinite ? value : 0
        return formatter.string(from: NSNumber(value: safe)) ?? "\(symbol)0.00"
    }

    /// Parse a free-form numeric string (the bill / tax input fields).
    /// Strips the currency symbol, grouping separators and whitespace.
    static func parse(_ text: String) -> Double {
        var cleaned = text
        if let groupingSeparator = formatter.groupingSeparator {
            cleaned = cleaned.replacingOccurrences(of: groupingSeparator, with: "")
        }
        cleaned = cleaned.replacingOccurrences(of: symbol, with: "")
        cleaned = cleaned.trimmingCharacters(in: .whitespacesAndNewlines)
        // Normalise the locale decimal separator to a dot for Double parsing.
        if let decimalSeparator = formatter.decimalSeparator, decimalSeparator != "." {
            cleaned = cleaned.replacingOccurrences(of: decimalSeparator, with: ".")
        }
        return Double(cleaned) ?? 0
    }
}

File: workspaces/ws_001_tipcalcdeluxe/android/app/src/main/java/com/appfactory/tipcalcdeluxe/model/CurrencyFormatter.kt
package com.appfactory.tipcalcdeluxe.model

import java.text.NumberFormat
import java.util.Locale

/**
 * Locale-aware currency formatting. The currency symbol is detected from the
 * device locale (spec feature: "Currency symbol auto-detection from locale").
 */
object CurrencyFormatter {

    private val format: NumberFormat by lazy {
        NumberFormat.getCurrencyInstance(Locale.getDefault()).apply {
            maximumFractionDigits = 2
            minimumFractionDigits = 2
        }
    }

    /** The currency symbol for the current locale (e.g. "$", "£", "AED"). */
    val symbol: String
        get() = runCatching { format.currency?.symbol }.getOrNull() ?: "$"

    /** Format a monetary value using the current locale's currency style. */
    fun string(value: Double): String {
        val safe = if (value.isFinite()) value else 0.0
        return runCatching { format.format(safe) }.getOrDefault("$symbol%.2f".format(safe))
    }

    /** Parse a free-form numeric string from a text field into a Double. */
    fun parse(text: String): Double {
        if (text.isBlank()) return 0.0
        var cleaned = text.trim()
        // Strip the currency symbol and any grouping separators.
        cleaned = cleaned.replace(symbol, "")
        cleaned = cleaned.replace(",", "")
        cleaned = cleaned.replace(" ", "")
        return cleaned.toDoubleOrNull() ?: 0.0
    }
}
